### PR TITLE
feature: send the recovery event when volume returns to the healthy state

### LINKED ISSUE #8: Send recovery events when volumes back to health
When volumes recover from unhealthy condition, check if abnormal events are sent, if so, send recovery events to indicate the volumes are healthy now.

### PR DESCRIPTION
**Pull request by fengzixu**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

1. Constructs an indexer of event for improving efficiency when we want to get historical abnormal events
2. Send the recovery event when the volume returns to the healthy state and abnormal event we sent before was found

**Which issue(s) this PR fixes**:

Fixes #8

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Send the recovery event when volume returns to the healthy state
```

### CODE REVIEW

File: pkg/csi-handler/pv_checker.go [hunk 1]
```diff
@@ -203,6 +219,46 @@ func (checker *PVHealthConditionChecker) CheckNodeVolumeStatus(kubeletRootPath s
...
+	pvcUID := string(pvc.ObjectMeta.GetUID())
+	key := fmt.Sprintf("%s:%s:%s", pvcUID, v1.EventTypeWarning, "VolumeConditionAbnormal")
+	events, err := checker.eventInformer.Informer().GetIndexer().ByIndex(util.DefaultEventIndexerName, key)
+	if err != nil {
+		klog.Warningf("Get abnormal event from indexer failed: %+v", err)
+		return nil
+	}
+
+	if len(events) > 0 {
+		checker.eventRecorder.Event(pvc, v1.EventTypeNormal, "VolumeConditionNormal", volumeCondition.GetMessage())
```
Comment (NickrenREN) [inline pkg/csi-handler/pv_checker.go:242]: Maybe we need to generate specific message to indicate this is a recovery event.

File: pkg/csi-handler/pv_checker.go [hunk 2]
```diff
@@ -203,6 +219,46 @@ func (checker *PVHealthConditionChecker) CheckNodeVolumeStatus(kubeletRootPath s
...
+	return nil
+}
+
+// sendRecoveryEventToPVC sends the recovery event to the pvc
+// If the volume condition is normal and abnormal event wasn't expired,
+// PVHealthConditionChecker should send recovery event.
+func (checker *PVHealthConditionChecker) sendRecoveryEventToPVC(pvc *v1.PersistentVolumeClaim, volumeCondition *VolumeConditionResult) error {
+	pvcUID := string(pvc.ObjectMeta.GetUID())
+	key := fmt.Sprintf("%s:%s:%s", pvcUID, v1.EventTypeWarning, "VolumeConditionAbnormal")
+	events, err := checker.eventInformer.Informer().GetIndexer().ByIndex(util.DefaultEventIndexerName, key)
```
Comment (NickrenREN) [inline pkg/csi-handler/pv_checker.go:235]: Just curious: will expired events be removed from event local cache right now after it is expired ?

File: pkg/csi-handler/pv_checker.go [hunk 3]
```diff
@@ -149,14 +161,18 @@ func (checker *PVHealthConditionChecker) CheckControllerVolumeStatus(pv *v1.Pers
...
+
 	// At the first stage, we just send PVC events
 	if volumeCondition.GetAbnormal() {
 		// Since pv status is bound, we believe PV controller, do not check pv.Spec.ClaimRef here.
-		pvc, err := checker.pvcLister.PersistentVolumeClaims(pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
-		if err != nil {
-			return err
-		}
 		checker.eventRecorder.Event(pvc, v1.EventTypeWarning, "VolumeConditionAbnormal", volumeCondition.GetMessage())
+	} else {
```
Comment (NickrenREN) [inline pkg/csi-handler/pv_checker.go:173]: What if error occurs when we execute API call ?
Default `abnormal` is True or False ?  IIUC, it is False
So we cann't directly send recovery events here.

File: pkg/csi-handler/pv_checker.go (hunk 1 above)
Comment (fengzixu) [reply pkg/csi-handler/pv_checker.go:242]: You mean the message part. Right?
An event usually includes

- type
- reason
- message

Is it ok to set `VolumeConditionNormal` as the reason for the recovery event?

File: pkg/csi-handler/pv_checker.go (hunk 3 above)
Comment (fengzixu) [reply pkg/csi-handler/pv_checker.go:173]: You mean `ControllerGetVolumeCondition` API was invoked failed. Right?
In that case, It should return from [here](https://github.com/kubernetes-csi/external-health-monitor/pull/36/files#diff-31795038ad9c7264079b907f4a04a702R161).

File: pkg/csi-handler/pv_checker.go (hunk 2 above)
Comment (fengzixu) [reply pkg/csi-handler/pv_checker.go:235]: I didn't check the implementation details. From my understanding, after one period of `event-ttl` passed, kube-apiserver will remove those events within `event-ttl`. In that case, it should trigger `DELETE Event` and handler of deletion should also be executed. So, the expired event in cache also should be removed.

File: pkg/csi-handler/pv_checker.go (hunk 3 above)
Comment (NickrenREN) [reply pkg/csi-handler/pv_checker.go:173]: Yes, right, i missed that part, thanks

File: pkg/csi-handler/pv_checker.go (hunk 1 above)
Comment (NickrenREN) [reply pkg/csi-handler/pv_checker.go:242]: Yes, and i am ok with `VolumeConditionNormal`

**Comment by fengzixu** [comment]
@xing-yang Any comment from you for this PR?

**Comment by xing-yang** [comment]
I took a quick look.  It looks fine.  I'll leave it to @NickrenREN to review and lgtm.

**Comment by xing-yang** [comment]
@NickrenREN any more comments on this PR?

**Comment by NickrenREN** [comment]
@xing-yang No, i will LGTM after it is updated

**Comment by NickrenREN** [comment]
btw: @fengzixu have you tested this ?

>I didn't check the implementation details for event resource. But I think it's the same with other native resources. From my understanding, after one period of event-ttl passed, kube-apiserver will remove those events within event-ttl. In that case, it should trigger DELETE Event and handler of deletion should also be executed. So, the expired event in cache also should be removed.

**Comment by fengzixu** [comment]
> btw: @fengzixu have you tested this ?
> 
> > I didn't check the implementation details for event resource. But I think it's the same with other native resources. From my understanding, after one period of event-ttl passed, kube-apiserver will remove those events within event-ttl. In that case, it should trigger DELETE Event and handler of deletion should also be executed. So, the expired event in cache also should be removed.

Not yet.

**Comment by fengzixu** [comment]
I added the testing logic as below in controller and set the `event-ttl` argument of kube-apiserver to 1m.
```go
// Run runs the volume health condition checking method
func (ctrl *PVMonitorController) Run(workers int, stopCh <-chan struct{}) {
	defer ctrl.pvQueue.ShutDown()

	klog.Infof("Starting CSI External PV Health Monitor Controller")
	defer klog.Infof("Shutting down CSI External PV Health Monitor Controller")

	if !cache.WaitForCacheSync(stopCh, ctrl.pvcListerSynced, ctrl.pvListerSynced, ctrl.podListerSynced) {
		klog.Errorf("Cannot sync cache")
		return
	}

	if ctrl.enableNodeWatcher {
		go ctrl.nodeWatcher.Run(stopCh)
	}

	pod, err := ctrl.podLister.Pods("test-event").Get("dnsutils")
	if err != nil {
		klog.Fatalf("failed to get pod : %v", err)
	}

	ctrl.eventRecorder.Event(pod, v1.EventTypeNormal, "test", "test")
	for {
		event, err := ctrl.eventsInformer.Lister().Events("test-event").List(labels.Everything())
		klog.Errorf("event error is: %+v", err)
		klog.Infof("Get the event: %+v", event)
		time.Sleep(5 * time.Second)
	}
```
The result illustrates the old event will be removed from the client-go cache after the `event-ttl` was passed.
```
I1012 16:58:48.915221       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:58:53.915602       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:58:53.915657       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:58:58.916011       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:58:58.916065       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:03.916400       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:03.916441       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:08.916862       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:08.916898       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:13.917218       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:13.917253       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:18.917528       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:18.917621       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:23.917976       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:23.918040       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:28.918318       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:28.918398       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:33.918721       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:33.918772       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:38.920675       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:38.920734       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:43.921003       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:43.921062       1 pv_monitor_controller.go:189] Get the event: [&Event{ObjectMeta:{dnsutils.163d4d3303b8926e  test-event /api/v1/namespaces/test-event/events/dnsutils.163d4d3303b8926e e7f7161b-e2fe-4720-8fc6-29b2e7d1fc46 20264 0 2020-10-12 16:58:43 +0000 UTC <nil> <nil> map[] map[] [] []  [{csi-external-health-monitor-controller Update v1 2020-10-12 16:58:43 +0000 UTC FieldsV1 {"f:count":{},"f:firstTimestamp":{},"f:involvedObject":{"f:apiVersion":{},"f:kind":{},"f:name":{},"f:namespace":{},"f:resourceVersion":{},"f:uid":{}},"f:lastTimestamp":{},"f:message":{},"f:reason":{},"f:source":{"f:component":{}},"f:type":{}}}]},InvolvedObject:ObjectReference{Kind:Pod,Namespace:test-event,Name:dnsutils,UID:60cd24f9-c8e8-421f-9356-f7ecc0415aa8,APIVersion:v1,ResourceVersion:19897,FieldPath:,},Reason:test,Message:test,Source:EventSource{Component:csi-pv-monitor-controller-io.kubernetes.storage.mock,Host:,},FirstTimestamp:2020-10-12 16:58:43 +0000 UTC,LastTimestamp:2020-10-12 16:58:43 +0000 UTC,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}]
E1012 16:59:48.921277       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:48.921299       1 pv_monitor_controller.go:189] Get the event: []
E1012 16:59:53.921426       1 pv_monitor_controller.go:188] event error is: <nil>
I1012 16:59:53.921448       1 pv_monitor_controller.go:189] Get the event: []
```

**Comment by fengzixu** [comment]
@xing-yang @NickrenREN Hi. I updated this PR. Please review it again.

**Comment by NickrenREN** [comment]
thanks @fengzixu 

/lgtm
/approve

**Pull request merged**